### PR TITLE
RD2:  Create schedule section LWC

### PR DESCRIPTION
**Pull request by RadSFDeveloper**
Extract RD2 entry form Schedule section (Donation Information) into a separate LWC so it can be used in the RD2 and Gift Entry forms.

# Critical Changes

# Changes

# Issues Closed

# Community Ideas Delivered

# New Metadata
Lightning Web Component:
- rd2EntryFormScheduleSection

# Deleted Metadata
###

### CODE REVIEW

**Comment by RadSFDeveloper** [comment]
**lurch: attach W-041485

**Comment by LurchTheButler** [comment]
Tracking <a href="https://foundation.my.salesforce.com/a2x1E000007haNEQAY" target="_blank">W-041485</a>

File: src/aura/RD2_EntryForm/RD2_EntryFormHelper.js [hunk 1]
```diff
@@ -10,39 +10,39 @@
...
+                    console.error(JSON.stringify(error));
+                }
+            });
     },
 
     /**
     * @description Handle events sent from the modal
     */
-    handleModalEvent: function(component, event) {
+    handleModalEvent: function (component, event) {
```
Comment (RadSFDeveloper) [inline src/aura/RD2_EntryForm/RD2_EntryFormHelper.js:37]: The VS Code formatter automatically updated files on save. Should we all use this formatter for LWC/LCs to avoid code review comments due to the coding style?

File: src/lwc/rd2EntryForm/rd2EntryForm.html [hunk 2]
```diff
@@ -1,15 +1,13 @@
...
         </h2>
     </header>
 
-    <lightning-record-edit-form record-id={recordId}
-                                object-api-name="npe03__Recurring_Donation__c"
-                                onsubmit={handleSubmit}
-                                onerror={handleError}
-                                onsuccess={handleSuccess}>
+    <lightning-record-edit-form record-id={recordId} object-api-name="npe03__Recurring_Donation__c"
+        onsubmit={handleSubmit} onerror={handleError} onsuccess={handleSuccess} data-id="outerRecordEditForm">
```
Comment (RadSFDeveloper) [inline src/lwc/rd2EntryForm/rd2EntryForm.html:10]: Added the `data-id` to clearly identify the outer record edit form.

File: src/lwc/rd2EntryForm/rd2EntryForm.html [hunk 3]
```diff
@@ -20,20 +18,18 @@ <h2 class="slds-text-heading_medium slds-text-heading--medium slds-hyphenate hea
                     </div>
                 </div>
             </template>
-            <template if:true={isFieldInfoReady}>
+            <template if:true={isRecordReady}>
```
Comment (RadSFDeveloper) [inline src/lwc/rd2EntryForm/rd2EntryForm.html:21]: It seems the `isRecordReady` check is enough to check when the form fields should be rendered.

File: src/lwc/rd2EntryForm/rd2EntryForm.html [hunk 4]
```diff
@@ -20,20 +18,18 @@ <h2 class="slds-text-heading_medium slds-text-heading--medium slds-hyphenate hea
...
+                        <c-util-page-level-message title={errorMessage.header} subtitle={errorMessage.detail}
+                            subtitle-class-override="slds-p-horizontal_medium slds-p-vertical_xx-small" variant="error">
                         </c-util-page-level-message>
                     </template>
-                
+
                     <div class="slds-section slds-is-open">
                         <h3 class="slds-section__title slds-theme_shade">
-                            <span class="slds-truncate slds-p-horizontal_small" title="Section Title">{customLabels.donorSectionHeader}</span>
+                            <span class="slds-truncate slds-p-horizontal_small">{customLabels.donorSectionHeader}</span>
```
Comment (RadSFDeveloper) [inline src/lwc/rd2EntryForm/rd2EntryForm.html:32]: Removed title since it is not required and has text that is not going to be translated.

File: src/lwc/rd2EntryForm/rd2EntryForm.html [hunk 5]
```diff
@@ -42,93 +38,36 @@ <h3 class="slds-section__title slds-theme_shade">
...
-                                </lightning-layout-item>
-
-                                <lightning-layout-item padding="horizontal-small" size="12"  small-device-size="6">
-                                    <lightning-input-field field-name={fields.paymentMethod.apiName}>
-                                    </lightning-input-field>
-                                </lightning-layout-item>
-                            </lightning-layout>
-                        </div>
-                    </div>
+                    <c-rd2-entry-form-schedule-section record-id={recordId}></c-rd2-entry-form-schedule-section>
```
Comment (RadSFDeveloper) [inline src/lwc/rd2EntryForm/rd2EntryForm.html:54]: New Schedule section (Donation Information) is self contained and will load fields data. The only param required is the `recordId`. Works for both new and existing RD.

File: src/lwc/rd2EntryForm/rd2EntryForm.js [hunk 6]
```diff
@@ -1,11 +1,11 @@
 import { LightningElement, api, track, wire } from 'lwc';
 import { fireEvent } from 'c/pubsubNoPageRef';
-import { ShowToastEvent } from 'lightning/platformShowToastEvent';
-
-import { getObjectInfo, getPicklistValues } from 'lightning/uiObjectInfoApi';
-import { getRecord} from 'lightning/uiRecordApi';
+import { showToast } from 'c/utilTemplateBuilder';
```
Comment (RadSFDeveloper) [inline src/lwc/rd2EntryForm/rd2EntryForm.js:3]: Using `c/utilTemplateBuilder` instead of creating the same `showToast` method in the RD2 Entry Form.

File: src/lwc/rd2EntryForm/rd2EntryForm.js [hunk 7]
```diff
@@ -40,120 +30,105 @@ import getSetting from '@salesforce/apex/RD2_entryFormController.getSetting';
...
+    /***
     * @description Dynamic render edit form CSS to show/hide the edit form 
     */
     get cssEditForm() {
-        return (!this.isLoading && this.isSettingReady && this.isFieldInfoReady && this.isRecordReady)
+        //Note: all of these flags need to be checked before the sections are displayed.
+        //If the isSettingReady is not checked, then the form on an error resets all fields
+        //including the Schedule section LWC fields.
+        //TODO: check why and how this expression can be simplifed.
+        return (!this.isLoading && this.isSettingReady && this.isRecordReady)
```
Comment (RadSFDeveloper) [inline src/lwc/rd2EntryForm/rd2EntryForm.js:69]: When these flags are not checked (especially the `isSettingReady`), the form *resets* field values on an error.

File: src/lwc/rd2EntryForm/rd2EntryForm.js [hunk 8]
```diff
@@ -40,120 +30,105 @@ import getSetting from '@salesforce/apex/RD2_entryFormController.getSetting';
...
+    connectedCallback() {
+        getSetting({ parentId: this.parentId })
+            .then(response => {
+                this.isAutoNamingEnabled = response.isAutoNamingEnabled;
+                this.handleParentIdType(response.parentSObjectType);
+            })
+            .catch((error) => {
+                this.handleError(error);
+            })
+            .finally(() => {
```
Comment (RadSFDeveloper) [inline src/lwc/rd2EntryForm/rd2EntryForm.js:86]: Let's use `finally` block to set vars common for both `then` and `catch`.

File: src/lwc/rd2EntryForm/rd2EntryForm.js [hunk 9]
```diff
@@ -40,120 +30,105 @@ import getSetting from '@salesforce/apex/RD2_entryFormController.getSetting';
...
     * @description Retrieve Recurring Donation Object info
     */
-    @wire(getObjectInfo, {objectApiName: RECURRING_DONATION_OBJECT.objectApiName })
-    wiredRDObjectInfo(response) {
+    @wire(getObjectInfo, { objectApiName: RECURRING_DONATION_OBJECT.objectApiName })
+    wiredRecurringDonationObjectInfo(response) {
         if (response.data) {
-            this.recurringDonationInfo = response.data;
-            this.setFields(this.recurringDonationInfo.fields);
+            let rdObjectInfo = response.data;
```
Comment (RadSFDeveloper) [inline src/lwc/rd2EntryForm/rd2EntryForm.js:113]: The `rdObjectInfo` is not used in the updated RD2 Entry Form, so it can be a local var.

File: src/lwc/rd2EntryForm/rd2EntryForm.js [hunk 10]
```diff
@@ -40,120 +30,105 @@ import getSetting from '@salesforce/apex/RD2_entryFormController.getSetting';
...
+
+            if (isNull(this.recordId)) {
+                this.isRecordReady = true;
+            }
         }
 
         if (response.error) {
-            this.isFieldInfoReady = true;
-            handleError(response.error);
+            this.isRecordReady = true;
```
Comment (RadSFDeveloper) [inline src/lwc/rd2EntryForm/rd2EntryForm.js:126]: Lesson learned: do not set `this.isRecordReady` outside of the `response.data` and `response.error` blocks. Moreover, errors are generated and displayed in the console log when the RD tab is displayed even before clicking on the New button.

File: src/lwc/rd2EntryForm/rd2EntryForm.js [hunk 11]
```diff
@@ -166,149 +141,113 @@ export default class rd2EntryForm extends LightningElement {
...
+    /***
     * @description Retrieve Recurring Donation record value
-    */ 
+    */
     @wire(getRecord, { recordId: '$recordId', fields: '$fieldInfos' })
     wiredRecurringDonationRecord(response) {
         if (response.data) {
             this.record = response.data;
-            this.modifyEditHeader();
+            this.header = editHeaderLabel + ' ' + this.record.fields.Name.value;
```
Comment (RadSFDeveloper) [inline src/lwc/rd2EntryForm/rd2EntryForm.js:174]: No need for a new method to replace one line of code.

File: src/lwc/rd2EntryForm/rd2EntryForm.js [hunk 12]
```diff
@@ -166,149 +141,113 @@ export default class rd2EntryForm extends LightningElement {
...
             this.isRecordReady = true;
         }
 
         if (response.error) {
             this.isRecordReady = true;
-            handleError(response.error);
+            this.handleError(response.error);
```
Comment (RadSFDeveloper) [inline src/lwc/rd2EntryForm/rd2EntryForm.js:180]: The current code in `feature/RD2` uses `handleError` methods without specifying `this.` in few places, so that can be corrected in another PR. @andrewyu-salesforce , @force2b
